ibis/common/tests/test_deferred.py

```diff
@@ -0,0 +1,519 @@
+        (">", "<", operator.gt),
+        (">=", "<=", operator.ge),
+    ],
+)
+def test_deferred_compare_operations(sym, rsym, op, table):
+    expr = op(_.a, _.b)
+    sol = op(table.a, table.b)
+    res = expr.resolve(table)
+    assert res == sol
+    assert repr(expr) == f"(_.a {sym} _.b)"
+
+    expr = op(1, _.a)
+    sol = op(1, table.a)
+    res = expr.resolve(table)
+    assert res == sol
+    assert repr(expr) == f"(_.a {rsym} 1)"
+
+
+@pytest.mark.parametrize(
+    "symbol, op",
+    [
+        ("-", operator.neg),
+        ("~", operator.invert),
+    ],
+)
+def test_deferred_unary_operations(symbol, op, table):
+    expr = op(_.a)
+    sol = op(table.a)
+    res = expr.resolve(table)
+    assert res == sol
+    assert repr(expr) == f"{symbol}_.a"
+
+
+@pytest.mark.parametrize("obj", [_, _.a, _.a.b[0]])
+def test_deferred_is_not_iterable(obj):
+    with pytest.raises(TypeError, match="object is not iterable"):
+        sorted(obj)
+
+    with pytest.raises(TypeError, match="object is not iterable"):
+        iter(obj)
+
+    with pytest.raises(TypeError, match="is not an iterator"):
+        next(obj)
+
+
+def test_deferrable(table):
+    @deferrable
+    def f(a, b, c=3):
+        return a + b + c
+
+    assert f(table.a, table.b) == table.a + table.b + 3
+    assert f(table.a, table.b, c=4) == table.a + table.b + 4
+
+    expr = f(_.a, _.b)
+    sol = table.a + table.b + 3
+    res = expr.resolve(table)
+    assert res == sol
+    assert repr(expr) == "f(_.a, _.b)"
+
+    expr = f(1, 2, c=_.a)
+    sol = 3 + table.a
+    res = expr.resolve(table)
+    assert res == sol
+    assert repr(expr) == "f(1, 2, c=_.a)"
+
+    with pytest.raises(TypeError, match="unknown"):
+        f(_.a, _.b, unknown=3)  # invalid calls caught at call time
+
+
+def test_deferrable_repr():
+    @deferrable(repr="<test>")
+    def myfunc(x):
+        return x + 1
+
+    assert repr(myfunc(_.a)) == "<test>"
+
+
+def test_deferred_set_raises():
+    with pytest.raises(TypeError, match="unhashable type"):
+        {_.a, _.b}  # noqa: B018
+
+
+@pytest.mark.parametrize(
+    "case",
+    [
+        param(lambda: ([1, _], [1, 2]), id="list"),
+        param(lambda: ((1, _), (1, 2)), id="tuple"),
+        param(lambda: ({"x": 1, "y": _}, {"x": 1, "y": 2}), id="dict"),
+        param(lambda: ({"x": 1, "y": [_, 3]}, {"x": 1, "y": [2, 3]}), id="nested"),
+    ],
+)
+def test_deferrable_nested_args(case):
+    arg, sol = case()
+
+    @deferrable
+    def identity(x):
+        return x
+
+    expr = identity(arg)
+    assert expr.resolve(2) == sol
+    assert identity(sol) is sol
+    assert repr(expr) == f"identity({arg!r})"
+
+
+# @pytest.mark.parametrize(
+#     "f, sol",
+#     [
+#         (lambda t: _.x + t.a, "(_.x + <column[int64]>)"),
+#         (lambda t: _.x + t.a.sum(), "(_.x + <scalar[int64]>)"),
+#         (lambda t: ibis.date(_.x, 2, t.a), "date(_.x, 2, <column[int64]>)"),
+#     ],
+# )
+# def test_deferred_repr_with_exprs(f, sol):
```

Comment:
There is no specialized `repr` for value expressions now, need to figure out how could we add one.

Reply:
@kszucs Is this resolved?

Reply:
Yes.